sling/frame/store.cc

```diff
@@ -833,6 +841,69 @@ SymbolDatum *Store::LocalSymbol(SymbolDatum *symbol) {
   return local;
 }
 
+uint64 Store::Fingerprint(Handle handle, uint64 seed) const {
+  // Fingerprint mixing constants.
+  enum FingerprintSeed : uint64 {
+    FP_NUMBER =  0xd1ac3c3a168f9a23,
+    FP_STRING =  0xedbf08a562d55ca0,
+    FP_FRAME =   0x07a535307e971126,
+    FP_SYMBOL =  0x06c498392bf66124,
+    FP_ARRAY =   0x7e71d2f093c19cd1,
+    FP_NIL =     0xe958f32bf433420c,
+    FP_INVALID = 0x159ba7c32c364f9b,
+  };
+
+  if (handle.IsNumber()) {
+    // Use the bit pattern of the integer or float for hashing.
+    return HashMix(HashMix(seed, FP_NUMBER), handle.bits);
+  } else {
+    if (handle.IsNil()) return HashMix(seed, FP_NIL);
+    const Datum *datum = GetObject(handle);
+    if (datum->IsFrame()) {
+      const FrameDatum *frame = datum->AsFrame();
+      if (frame->IsNamed()) {
+        // Use the (first) frame id for hashing.
+        Handle id = frame->get(Handle::id());
+        DCHECK(!id.IsNil());
+        const SymbolDatum *symbol = GetSymbol(id);
+        return HashMix(seed, symbol->hash.bits);
```

Comment:
I did it this way to avoid mixing in FP_FRAME too often and to make this case a little faster since this is a very common case in the recursion. I did mix in FP_SYMBOL for pure symbols to make this different than the fingerprints for the corresponding frame, but this is a pretty rare case.